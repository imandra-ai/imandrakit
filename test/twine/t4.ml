type t1 =
  | F of (string[@as_bytes])
  | G of string
[@@deriving show, twine]

let x = F "hello\x000world"
let enc_x = Imandrakit_twine.Encode.to_string t1_to_twine x;;

Format.printf "x: %a@." pp_t1 x;;
Format.printf "enc(x): %S@." enc_x;;
Format.printf "len: %d@." (String.length enc_x);;
Format.printf "hex:@.%s@." (Hex.hexdump_s @@ Hex.of_string enc_x)

let x' = Imandrakit_twine.Decode.decode_string t1_of_twine enc_x;;

Format.printf "deser: %a@." pp_t1 x'

type t2 = { t1s: t1 option list } [@@deriving show, twine]

let t2 = { t1s = [ Some x; None; Some x ] }
let enc_t2 = Imandrakit_twine.Encode.to_string t2_to_twine t2;;

Format.printf "t2: %a@." pp_t2 t2;;
Format.printf "enc(t2): %S@." enc_t2;;
Format.printf "len: %d@." (String.length enc_t2);;
Format.printf "hex:@.%s@." (Hex.hexdump_s @@ Hex.of_string enc_t2)

let t2' = Imandrakit_twine.Decode.decode_string t2_of_twine enc_t2;;

Format.printf "deser: %a@." pp_t2 t2'
