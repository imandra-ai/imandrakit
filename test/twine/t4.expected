x: (T4.F "hello\0000world")
enc(x): "\176Lhello\0000world\r"
len: 15
hex:
00000000: b04c 6865 6c6c 6f00 3077 6f72 6c64 0d    .Lhello.0world.

deser: (T4.F "hello\0000world")
t2: { T4.t1s =
      [(Some (T4.F "hello\0000world")); None; (Some (T4.F "hello\0000world"))
        ]
      }
enc(t2): "\176Lhello\0000worlda\254\176Lhello\0000worlda\254c\255\003\002\245a\245\001"
len: 40
hex:
00000000: b04c 6865 6c6c 6f00 3077 6f72 6c64 61fe  .Lhello.0worlda.
00000001: b04c 6865 6c6c 6f00 3077 6f72 6c64 61fe  .Lhello.0worlda.
00000002: 63ff 0302 f561 f501                      c....a..

dump:
[0x0]: C_0("hello\0000world")
[0xe]: [@0x0] (len=1)
[0x10]: C_0("hello\0000world")
[0x1e]: [@0x10] (len=1)
[0x20]: [@0xe, null, @0x1e] (len=3)
[0x25]: [@0x20] (len=1)

deser: { T4.t1s =
         [(Some (T4.F "hello\0000world")); None;
           (Some (T4.F "hello\0000world"))]
         }
